rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,406 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 136 - Modern Signature Algorithms
+
+## Required Approvers
+
+TODO
+
+## What
+
+Teleport should support modern key types and signature algorithms, currently
+only RSA2048 keys are supported with the PKCS#1 v1.5 signature scheme.
+This applies to CA keys and client (user/host) keys, but each can/will be
+addressed individually.
+
+## Why
+
+Modern algorithms like ECDSA and Ed25519 offer better security properties with
+smaller keys that are faster to generate and sign with.
+Some of the more restrictive security policies are starting to reject RSA2048
+(e.g. [RHEL 8's FUTURE policy](https://access.redhat.com/articles/3642912)).
+
+## Details
+
+### Algorithms
+
+These algorithms are being considered for support:
+
+#### RSA
+
+Private key sizes: 2048, 3072, 4096
+
+Signature algorithms: PKCS#1 v1.5, (maybe PSS for TLS and JWT? SSH does not support it)
+
+Digest/hash algorithms: SHA512 for SSH, SHA256 for TLS
+
+Considerations:
+
+* RSA2048 is the current default and deviating from it by default may break
+  compatibility with third-party components and protocols
+* RSA has the most widespread support among all protocols
+* Certain database protocols only support RSA client certs
+  * <https://docs.snowflake.com/en/user-guide/key-pair-auth#step-2-generate-a-public-key>
+* If we must continue to support RSA, we might as well support larger key sizes,
+  3072 and 4096-bit are the most commonly used and supported by e.g. GCP KMS.
+* golang.org/x/crypto/ssh always uses SHA512 hash with RSA public keys
+  <https://github.com/golang/crypto/blob/0ff60057bbafb685e9f9a97af5261f484f8283d1/ssh/certs.go#L443-L445>
+* crypto/x509 always uses SHA256 hash with RSA public keys
+  <https://github.com/golang/go/blob/dbf9bf2c39116f1330002ebba8f8870b96645d87/src/crypto/x509/x509.go#L1411-L1414>
+* ssh only supports the PKCS#1 v1.5 signature scheme
+  <https://datatracker.ietf.org/doc/html/rfc8332>
+* FIPS 186-5 approves all listed options
+* BoringCrypto supports all listed options
+
+#### ECDSA
+
+Curves: P-256, (maybe P-384, P-521?)
+
+Digest/hash algorithms: SHA256, (or SHA384, SHA512 for P-384, P-521)
+
+Considerations:
+
+* ECDSA has good support across SSH and TLS protocols for both client and CA
+  certs.
+* ECDSA key generation is *much* faster than RSA key generation.
+* ECDSA signatures are faster than RSA signatures.
+* FIPS 186-5 approves all listed options
+* BoringCrypto supports all listed options
+
+#### EdDSA
+
+Curves: Ed25519 (this is the only curve supported in Go)
```

Comment:
@Joerger I remember you did some research about it.

Reply:
Actually yes, I think [this issue](https://github.com/gravitational/teleport-private/issues/299) is related. I've added more details in that issue. 

tldr; separating the SSH and TLS keys is probably necessary, regardless of the compatibility concerns.

Reply:
In the latest revision of this RFD, all certs will have their own unique private key